wolfSSL pkg addition with examples

danielinux opened the pull request:
### Contribution description

This patch superseeds PR #9894  and contains wolfSSL support in a RIOT pkg.

This patch includes:
  - independent modules that can be selected and integrated separately into applications
  - single cryptography algorithm selectable as separate modules
  - Support for TLS over TCP/IP using **sockets**
  - Support for DTLS over UDP/IP using **sock_udp** interface

### Modules
**wolfCrypt**:
 - wolfcrypt_poly1305
 - wolfcrypt_chacha
 - wolfcrypt_ed25519
 - wolfcrypt_aes
 - wolfcrypt_pwdbased
 - wolfcrypt_asn
 - wolfcrypt_random
 - wolfcrypt_* (e.g. ecc, rsa, sha3, ...)

**wolfSSL**:
 - wolfssl_tls
 - wolfsll_tls13
 - wolfssl_dtls
 - wolfssl_* (e.g. ocsp, crl, ...)


### Testing procedure: sock_udp

The example `dtls-wolfssl` provided uses similar shell commands to those in dtls-echo:

`dtlss`
 start dtls server on port 11111

`dtlsc addr`
start dtls client demo connection to server at address *addr*

`wolftest`
 run all sanity checks to ensure that the crypto engine is working on the target

#### Compile and run dtls-wolfssl

wolfSSL works on any 32-bit target with support for `gnrc_networking`. 

The test procedure for native riot-to-riot test is the following:

#### Prepare the bridge interface linking two tuntap:

    ./../../dist/tools/tapsetup/tapsetup --create 2

#### Run the server
```
    $ make all; PORT=tap1 make term
    > ifconfig
```
*copy the server address*

```
   > dtlss
```
#### Run the client
```
    $ PORT=tap0 make term
    > dtlsc <IPv6's server address[%netif]>
```

### Testing procedure: full TCP/IP TLS demo

The examples `wolfssl-client` and `wolfssl-server` provide a TLS demo with full TCP/IP connectivity, and supports all 32-bit platforms, used in combination with the lwIP package. The TCP socket support depends on `lwip_sock_tcp`.

### Issues/PRs references

This PR includes all previous work made in #9894 and #7348, rebased as single commit on the current master.

The wolfSSL library package temporarly lives in the feature branch https://github.com/danielinux/wolfssl/tree/riot-os until the new callbacks for GNRC are merged in.

### Acknowledgements

Thanks to @cladmi for reviewing the modules integration and helping with modularization of the single components.

Thanks to @bergzand for testing wolfcrypt_ed25519 and suggesting code-size optimization settings.

Please consider for inclusion.

Any feedback is welcome.

-- @danielinux -- @kaleb-himes -- @wolfssl

jcarrano commented:
@danielinux I'm rushing to finish #9565 so that you can use it to embed the certificates.

danielinux commented:
I made the CI robots happy by fixing a few issues in the code and in the markdown files.

Please let me know if there are any further actions needed to speed up the review of this PR.

Thanks,

-- @danielinux -- @wolfssl

MrKevinWeiss commented:
release status ping, should I change the milestone?

I think @MichelRottleuthner has something to say

MrKevinWeiss commented:
Something tells me this will continue to be pushed to the future release.

danielinux commented:
It would be nice to hear any comment about this PR.

wolfSSL's support original PR (#7348) is almost two years old while typing this. Since then, we have widely improved the integration. Now we have a proper pkg, we build single wolfcrypt/wolfssl components into separate modules, and we support both posix and GNRC socket interfaces.

The latest version of this code is a few weeks old and is based on wolfSSL 4.0.0-stable.

I have updated this branch regularly in the past months, to sync up to latest RIOT and wolfSSL master branches and test/fix the examples.

MrKevinWeiss commented:
I don't think it will get in this release but I can bug @MichelRottleuthner to help with the review

MichelRottleuthner commented:
I finally found some time to spend on this PR and I'm willing to help wherever I can to get this merged.
First: I feel really sorry for how long this PR is already waiting to get merged, especially considering how valuable it is. I know how frustrating it can be when it takes that long to get a feature in.
And I hope it didn't discourage you to to further contribute to RIOT.
There are a lot of people in the community that want this feature in - but sadly even more people that are short on time ;)
Anyway, thank you a lot for staying active!

That being said, let's move on to the technical stuff.
It's very much appreciated that you separated and integrated everything as individual modules.
Looking at the history of all the related PRs clearly shows a huge improvement.

I didn't do a detailed code review yet but when quickly browsing thru the code I couldn't spot major problems.
The PR comes with quite a lot of different examples/tests:
under examples we have:
`dtls-wolfssl`, `wolfssl-client`, `wolfssl-server`
under tests we have:
`wolfcrypt-ed25519-verify`, `wolfssl_crypto_benchmark`, `wolfssl_crypto_test`

While I generally like to have more than one or two examples/tests, I think it adds unnecessary complexity for getting a first version of a new feature in.
`tests/wolfssl_crypto_test` could basically be removed as this is already contained as `wolftest` command in `examples/dtls-wolfssl`, right?
Would it be possible to integrate the other tests i.e. `wolfcrypt-ed25519-verify`, `wolfssl_crypto_benchmark` in a similar way?

When trying to build `examples/wolfssl-client` or `examples/wolfssl-server` I get the following compilation error:
```
$ make
Building application "wolf_server" for "native" with MCU "native".
"make" -C /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src
In file included from /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/wolfssl/ssl.h:134,
                 from /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/wolfssl/internal.h:29,
                 from /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/wolfio.c:37:
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/wolfio.c: In function ‘wolfIO_Recv’:
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/wolfssl/wolfio.h:274:27: error: implicit declaration of function ‘recv’ [-Werror=implicit-function-declaration]
  274 |     #define RECV_FUNCTION recv
      |                           ^~~~
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/wolfio.c:624:18: note: in expansion of macro ‘RECV_FUNCTION’
  624 |     recvd = (int)RECV_FUNCTION(sd, buf, sz, rdFlags);
      |                  ^~~~~~~~~~~~~
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/wolfio.c: In function ‘wolfIO_Send’:
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/wolfssl/wolfio.h:273:27: error: implicit declaration of function ‘send’ [-Werror=implicit-function-declaration]
  273 |     #define SEND_FUNCTION send
      |                           ^~~~
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/wolfio.c:634:17: note: in expansion of macro ‘SEND_FUNCTION’
  634 |     sent = (int)SEND_FUNCTION(sd, buf, sz, wrFlags);
      |                 ^~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [/home/michel/devel/wolfssl/RIOT/Makefile.base:83: /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/native/wolfssl/wolfio.o] Error 1
make[1]: *** [/home/michel/devel/wolfssl/RIOT/Makefile.base:20: ALL--/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src] Error 2
make: *** [/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/../../Makefile.include:485: /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/native/application_wolf_server.a] Error 2
```
[The system setup used.](https://gist.github.com/MichelRottleuthner/28e9ed7e1ca216f124bc747280936494)
So for now I didn't consider these applications for further testing and as of now focused on running `examples/dtls-wolfssl` on variuos platforms.

Here is a first overview of what I found working on which platform:

| *board*       | *wolftest* | *dtlss*  | *dtlsc*    |
|---|---|---|---|
| native        | PASSED     | PASSED   | PASSED     |
| samr21-xpro   | PASSED     | ?        | FAILED (2) |
| nucleo-f303re | PASSED     | FAILED   | PASSED     |
| nucleo-f207zg | PASSED     | PASSED   | ?          |
| nucleo-f091rc | FAILED (1) | FAILED   | FAILED (2) |
| nucleo-f767zi | PASSED     | PASSED   | PASSED     |
| nucleo-l476rg | PASSED     | PASSED   | PASSED     |

The failed ones under dtlss didn't actually print an error on the server side but made the client fail to connect.
For the ones marked with '?' I couldn't exactly tell which side was inducing the problems, I will investigate further.
Details on how the nodes failed for the ones that have a number:

1)
```
...
RANDOM   test passed!
ecc_test_curve_size 32 failed!: -125
ECC      test failed!
 error = -125
#! exit -1: powering off
```

2)
```
...
connecting to server...
Error allocating ssl session
```

Do you have any hints on what I could try to further investigate?
Could this be related to stacksize being to small or maybe timeouts that are too tight for rather slow platforms?

Some low priority comments regarding the shell commands of dtls-wolfssl:
is it expected/required that the system stops after running the wolftest command?
At least I didn't expect that and its not mentioned in the command description.
Also, the `dtlss` description says that it can be used to start and stop a server, but actually it just starts it and then auto-closes after fist connection.

danielinux commented:
I understand how difficult it can be to find time to review these PR, so no worries!
Thanks for your time, I'll try to address the issues here.

> It's very much appreciated that you separated and integrated everything as individual modules.
> Looking at the history of all the related PRs clearly shows a huge improvement.

Kudos to @cladmi for the patience guiding me through this part.

> The PR comes with quite a lot of different examples/tests:
> under examples we have:
> dtls-wolfssl, wolfssl-client, wolfssl-server

WolfSSL support can integrate with either GNRC (via wolfssl + gnrc_sock_udp/tcp) or POSIX socket calls (via wolfssl_socket + posix_socket modules).

`dtls-wolfssl` is an example using GNRC sockets. wolfssl-[client|server] are
the older demos, that depend on POSIX-compliant sockets.

I agree we could remove these examples if they are not needed/less interesting,
but on the other hand they are the only example that use wolfssl_socket module (with
posix_socket).

I've checked the problem with compiling these tests, it's due to a missing
include, I'll push a fix.

> under tests we have:
> wolfcrypt-ed25519-verify, wolfssl_crypto_benchmark, wolfssl_crypto_test

These are wolfcrypt-only test/benchmark units, that have also been used to
evaluate performance of single in other contexts (see @bergzand's work on
secure boot).

On some platforms, I expect that there might not be enough resources (code
size/heap/stack) to run full SSL (e.g. DTLS), while single wolfcrypt
algorithms (that can now be selected as separate modules) might be the right fit
for specific purposes on constrained targets. For this reason I think it might be useful
to keep some of the 'tests'.

> When trying to build examples/wolfssl-client or examples/wolfssl-server I get the following compilation error:

I'll push a fix for this.

> So for now I didn't consider these applications for further testing and as of now focused on running examples/dtls-wolfssl on variuos platforms.

> Here is a first overview of what I found working on which platform:
[...]


> The failed ones under dtlss didn't actually print an error on the server side but made the client fail to connect.
> For the ones marked with '?' I couldn't exactly tell which side was inducing the problems, I will investigate further.
> Details on how the nodes failed for the ones that have a number:

> RANDOM   test passed!
> ecc_test_curve_size 32 failed!: -125

Error -125 is 'out of memory error'. The target could not provide the memory
required to perform the ecc_test.

> connecting to server...
> Error allocating ssl session

Sounds like a memory problem again, wolfSSL_new is failing to create the session
object (malloc failed?).

> Do you have any hints on what I could try to further investigate?
> Could this be related to stacksize being to small or maybe timeouts that are too tight for rather slow platforms?

Stack (or more likely heap) might be a problem initializing large objects (TLS session) on some platforms. Wolfcrypt-only tests have more chances to run on those.

Perhaps I should add targets to add targets to BOARD_INSUFFICIENT_MEMORY depending
on the module selection. Feel free to suggest a solution.

> Some low priority comments regarding the shell commands of dtls-wolfssl:

I took notes about these, I will fix the dtls example behavior.

MichelRottleuthner commented on examples/dtls-wolfssl/dtls-server.c:
```diff
@@ -0,0 +1,135 @@
+extern unsigned int server_key_len;
+
+static sock_tls_t skv;
+static sock_tls_t *sk = &skv;
```
minor: why the additional pointer?

danielinux replied:
just for readability, to avoid extra-dereferences in the code below.

MichelRottleuthner commented on examples/dtls-wolfssl/main.c:
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ * Copyright (C) 2018 Inria
```
please update

danielinux replied:
updated

MichelRottleuthner commented on examples/wolfssl_client/main.c:
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2006-2017 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL.
```
The copyright note should be adapted to the same format as other RIOT files

danielinux replied:
updated all notices

MichelRottleuthner commented on examples/wolfssl_server/Makefile:
```diff
@@ -0,0 +1,57 @@
+
+#DISABLE_MODULE += auto_init
+
+#CFLAGS += ??
```
Why are these lines here?

MichelRottleuthner commented on examples/wolfssl_server/main.c:
```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2006-2017 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL.
```
same as with the copyright notes before

danielinux replied:
done

MichelRottleuthner commented on examples/wolfssl_server/main.c:
```diff
@@ -0,0 +1,189 @@
+    struct sockaddr_in servAddr;
+    struct sockaddr_in clientAddr;
+    socklen_t          size = sizeof(clientAddr);
+    char               buff[256];
```
magic number

MichelRottleuthner commented on pkg/wolfssl/Makefile:
```diff
@@ -0,0 +1,17 @@
+PKG_NAME=wolfssl
+PKG_URL=https://github.com/danielinux/wolfssl.git
+#PKG_URL=https://github.com/wolfssl/wolfssl.git
```
is this meant to be changed to the upstream version for the final integration?

danielinux replied:
Yes, and since wolfssl PR has been accepted last week, I've updated this so it finally points to the official masterbranch

MichelRottleuthner commented on pkg/wolfssl/doc.txt:
```diff
@@ -0,0 +1,184 @@
+/**
```
this doc file probably needs an update: the dtls-wolfssl application is not listed here. (and maybe some other applications like the wolfssl_crypto_test will be removed)

danielinux replied:
Doc updated.

MichelRottleuthner commented on pkg/wolfssl/include/user_settings.h:
```diff
@@ -0,0 +1,308 @@
+#else
+#   define WOLFSSL_GENERAL_ALIGNMENT 4
+#   ifdef CPU_ARM
+#      define TFM_ARM
```
I don't know what this actually does, but would this be still valid for non-ARM CPUs? (same for the alignment)

danielinux replied:
The port is 32-bit only. The ARM specific flags are used to enable platform-specific optimizations, only if available (i.e. when CPU_ARM is defined globally)

MichelRottleuthner replied:
I see. So we should exclude all non-32 bit platforms anyway? And for native it is also intended that no value is provided? (asking because the define for native is commented out)

MichelRottleuthner commented on pkg/wolfssl/include/user_settings.h:
```diff
@@ -0,0 +1,308 @@
+#	define NO_HMAC
+#endif
+
+#undef NO_SHA
```
some of the configurations here have a comment before (e.g. HMAC above), some not. IMO the comments are not really adding information so I'd vote for removing them. But in any case try keeping it consistent.

danielinux replied:
comments removed

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.c:
```diff
@@ -0,0 +1,119 @@
+/* Copyright (C) 2014 Freie Universität Berlin
```
please adapt

danielinux replied:
done

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.c:
```diff
@@ -0,0 +1,119 @@
+#include <wolfssl/ssl.h>
+
+#define MODE_TLS 0
+#define MODE_DTLS 1
```
these two defines are not acutally used in this file, right?

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,339 @@
+#define MODE_DTLS 1
+
+#define MODE_TLS 0
+#define MODE_DTLS 1
```
double copy paste? ;)

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,339 @@
+ *          ((local->netif != SOCK_ADDR_ANY_NETIF) ||
+ *          (remote->netif != SOCK_ADDR_ANY_NETIF))` if neither is `NULL`).
+ * @return  -ENOMEM, if not enough resources can be provided for `sock` to be
+ *          created, 0 upon success.
```
"0 upon success" already listed in the first return description

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,339 @@
+ * @pre `(addr != NULL)`
+ *
+ * @param[in] sock     The resulting sock object.
+ * @param[in] addr      Remote end point for the DTLS session.
```
minor: indent of the two description lines

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,339 @@
+ * @pre `(sk != NULL)`
+ * @pre `(addr != NULL)`
+ *
+ * @param[in] sock     The resulting sock object.
```
the param is actually named "sk" here

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,339 @@
+ * @return  -ENOMEM, if not enough resources can be provided for the session to be
+ *          created
+ */
+int sock_dtls_session_create(sock_tls_t *sk);
```
param name "sk" here, "sock" in the description

MichelRottleuthner commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,339 @@
+
+
+#ifdef MODULE_SOCK_TCP
+/* TODO */
```
Shouldn't this trigger some kind of build error, as its currently not supported? Maybe add a little bit more information.on why this is not available.

MichelRottleuthner commented on tests/wolfcrypt-ed25519-verify/main.c:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
```
please adapt

danielinux replied:
done

MichelRottleuthner commented on tests/wolfcrypt-ed25519-verify/main.c:
```diff
@@ -0,0 +1,91 @@
+ * @file
+ * @brief       ed25519 signature verification application
+ *
+ * @author      wolfSSL <info@wolfssl.com>
```
can you replace that by the actual person?

danielinux replied:
fixed

MichelRottleuthner commented on tests/wolfcrypt-ed25519-verify/tools/ed25519_sign_msg.c:
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
```
please adapt

MichelRottleuthner commented on tests/wolfssl_crypto_benchmark/Makefile:
```diff
@@ -0,0 +1,48 @@
+BOARD ?= samd21-xpro #Currently missing implementation for strncasecmp
+#BOARD ?= pic32-wifire
+#BOARD ?= frdm-k64f
+#BOARD ?= stm32f4discovery
```
please remove the boards that are commented out. If you think its is important to note that it was tested on those boards you could just add a comment.

MichelRottleuthner commented on tests/wolfssl_crypto_benchmark/Makefile:
```diff
@@ -0,0 +1,48 @@
+
+# BOARD_WHITELIST := To Be Determined
+# BOARD_BLACKLIST := To Be Determined
+# BOARD_INSUFFICIENT_MEMORY := To Be Determined
```
obviously, but maybe its about time ;)

MichelRottleuthner commented on tests/wolfssl_crypto_benchmark/Makefile:
```diff
@@ -0,0 +1,48 @@
+# Also need third-party string library if attempting to use ecc that will
+# implement strncasecomp
+#    USEPKG += lwip
+#    export CFLAGS += "-DHAVE_LWIP_NATIVE=1"
```
if it is recommended to use this configuration, why is it commented out?

danielinux replied:
removed

MichelRottleuthner commented on tests/wolfssl_crypto_benchmark/main.c:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2006-2017 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL.
```
as with the files before, please adapt to our default style for copyright notes.

danielinux replied:
reviewed all copyright notes in all the contributed files

MichelRottleuthner commented on tests/wolfssl_crypto_test/Makefile:
```diff
@@ -0,0 +1,47 @@
+#BOARD ?= native
+#BOARD ?= samd21-xpro #Currently missing implementation for strncasecmp
+#BOARD ?= pic32-wifire
+#BOARD ?= frdm-k64f
```
same as above

danielinux replied:
removed

MichelRottleuthner commented on tests/wolfssl_crypto_test/Makefile:
```diff
@@ -0,0 +1,47 @@
+
+# BOARD_WHITELIST := To Be Determined
+# BOARD_BLACKLIST := To Be Determined
+# BOARD_INSUFFICIENT_MEMORY := To Be Determined
```
same as above

danielinux replied:
this is also fixed (see new blacklist in the unified test app)

MichelRottleuthner commented on tests/wolfssl_crypto_test/Makefile:
```diff
@@ -0,0 +1,47 @@
+QUIET ?= 1
+
+# May need these for seeding wolfCrypt DRBG
+#USEMODULE += prng_minstd
```
The crypto_benchmark above also listed `#USEMODULE += random` why not the same here?

danielinux replied:
this should be gone now, and the two tests (wolcrypt_test + benchmark) have been unified to reduce scattering in the test directory.

MichelRottleuthner commented on tests/wolfssl_crypto_test/main.c:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2006-2017 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL.
```
same as before

danielinux replied:
fixed everywhere

MichelRottleuthner commented:
>Stack (or more likely heap) might be a problem initializing large objects (TLS session) on some >platforms. Wolfcrypt-only tests have more chances to run on those.

Ok. IIRC, wolfssl also has some option to work with static memory - isn't that used for RIOT? If it is used, I'd actually expect that the build would fail with some "region whatever overflowed" message for too constrained devices. If not, what needs to be done to use static memory?


>Perhaps I should add targets to add targets to BOARD_INSUFFICIENT_MEMORY depending
>on the module selection. Feel free to suggest a solution.

If there is a way to let it fail on build we can use `make buildtest` to derive a list. I think setting BOARD_INSUFFICIENT_MEMORY based on the selected modules isn't really required. It should just be ensured that the list is valid for the default application configuration. The CI wouldn't enable disable modules on its own, and a dev who changes the configuration would still have to check if everything works as expected. In any case using (big enough) static memory would help by just failing to build if it can not be ensured to work.

I just tested (only compilation for now) of the wolfssl_client/server applications with your fix. Building for an arm-none-eabi-gcc target (nucleo-l476rg) failed with implicit declaration of `strnlen`. Looking at the code it should be safe to just replace with strlen() because in both cases the string is not user-provided, and also it is in a non-critical part of an example application. On the wolfssl_client application I also got an "undefined reference to `inet_pton'" error that could be fixed by adding "USEMODULE += posix_inet" to the Makefile.

regarding wolfcrypt-ed25519-verify, wolfssl_crypto_benchmark, wolfssl_crypto_test:
>(...) These are wolfcrypt-only test/benchmark units (...)

I wonder if the binary size grows too much if those get combined into a singe application providing three test commands?

MichelRottleuthner commented:
Thanks for adding the fixes. I tried again with `examples/wolfssl_server` and `examples/wolfssl_client`. 
The server crashes for me on native:
```
/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/native/wolf_server.elf tap1 
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2019.07-devel-607-gcd5c1-prtest_10308_wolfssl)
This is the wolfSSL Server!
Server is running on 127.0.0.1 and listening on port 11111
make: *** [/home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/../../Makefile.include:558: term] Segmentation fault (core dumped)
```
Coming from the main the crash happens in the `wolfSSL_CTX_use_PrivateKey_buffer` function.
Here is a gdb backtrace of the crash:
```
#0  0xf7d61ad4 in free () from /usr/lib32/libc.so.6
#1  0x5655911f in free (ptr=0xfaf414a4) at /home/michel/devel/wolfssl/RIOT/cpu/native/syscalls.c:175
#2  0x5657f894 in wc_RsaCleanup (key=0x565c3130) at /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/wolfcrypt/src/rsa.c:241
#3  0x5657fa69 in wc_FreeRsaKey (key=0x565c3130) at /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/wolfcrypt/src/rsa.c:520
#4  0x5655f88b in ProcessBuffer (ctx=0x565c2170, buff=0x5659e5a0 <server_key_der_2048> "0\202\004\245\002\001", sz=1193, format=2, type=1, ssl=0x0, used=0x0, userChain=0)
    at /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/ssl.c:4929
#5  0x56560f1c in wolfSSL_CTX_use_PrivateKey_buffer (ctx=0x565c2170, in=0x5659e5a0 <server_key_der_2048> "0\202\004\245\002\001", sz=1193, format=2)
    at /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/bin/pkg/native/wolfssl/src/ssl.c:11284
#6  0x56558065 in main () at /home/michel/devel/wolfssl/RIOT/examples/wolfssl_server/main.c:117
```

For arm-none-eabi targets I can't compile. Removing strnlen or adding `#define _DEFAULT_SOURCE 1` at the top of main.c could be used to solve this. 

To actually enable radio support on platforms that come with a radio `USEMODULE += netdev_default` should be added to the Makefile. For 802.15.4 radios this wont work with the default `LWIP_IPV4 = 1` setting. Though, overwriting actually doesn't make a difference for the application as values like the IP are hardcoded in v4 style and calls to `socket(AF_INET, ...)` etc. all use `..INET` instead of `..INET6`.
Did you try these examples on an MCU platform? Because if these examples will only work on native anyway, I don't think we should keep them. Which MCU-platform did you use for running these examples?

MichelRottleuthner commented:
I gave the `tests/wolfssl_crypto_benchmark` a try.
Out-of-the-box I didn't manage to run it on any MCU platform until I figured out that the default benchmark configuration is just too heavy. Adding `#define BENCH_EMBEDDED` to `pkg/wolfssl/include/user_settings.h` solved that. This should be done automatically if compiling for an MCU target.

Additionally I needed to add `USEMODULE +=  xtimer` to stop the benchmark from freezing and `USEMODULE +=  printf_float` to properly print the results.
We should figure out if these dependencies should be added to any of the submodules of the wolfssl package or if it is just a dependency of this benchmark application. Depending on that they should be added to the respective Makefile.dep or the app's Makefile.

danielinux commented on examples/dtls-wolfssl/cert.c:
```diff
@@ -0,0 +1,83 @@
+  0x25, 0xd4, 0xb0, 0xc9, 0x25, 0x7d, 0xca, 0x7a, 0x5d, 0xba, 0xc4, 0xb2,
+  0xf6, 0x7d, 0x04, 0xc7, 0xbd, 0x62, 0xc9, 0x20
+};
+unsigned int server_cert_len = 788;
```
this file can be automatically generated by `xxd -i` and the size is appended automatically.

MichelRottleuthner replied:
I see. Then just leave it as is

danielinux commented on examples/dtls-wolfssl/cert.c:
```diff
@@ -0,0 +1,83 @@
+  0x11, 0x02, 0x34, 0x86, 0xe8, 0x20, 0x58, 0x33, 0x0b, 0x80, 0x34, 0x89,
+  0xd8
+};
+unsigned int server_key_len = 121;
```
see previous comment

danielinux commented on examples/dtls-wolfssl/Makefile:
```diff
@@ -0,0 +1,57 @@
+USEMODULE += wolfcrypt_ecc
+USEMODULE += wolfssl_dtls
+
+#CFLAGS+=-g -ggdb3
```
Added a comment

danielinux commented on examples/dtls-wolfssl/Makefile:
```diff
@@ -0,0 +1,57 @@
+CFLAGS += -DDEVELHELP
+
+# Change this to 0 show compiler invocation lines by default:
+QUIET ?= 0
```
fixed

danielinux commented on examples/dtls-wolfssl/README.md:
```diff
@@ -0,0 +1,43 @@
+## Boards
+
+Boards that do not support the `../gnrc_networking` example are included
+in the `BOARD_INSUFFICIENT_MEMORY`, plus the board `cc2650stk`.
```
reworded

danielinux commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,128 @@
+#define SERVER_PORT 11111
+
+extern const unsigned char server_cert[788];
+extern const unsigned long server_cert_len;
```
fixed

danielinux commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,128 @@
+    sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+    sock_udp_ep_t remote = SOCK_IPV6_EP_ANY;
+
+    /* Parsing <address>[:<iface>]:Port */
```
fixed comment

danielinux commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,128 @@
+        return -1;
+    }
+    if (argc != 2) {
+        printf("Usage: %s <server-address>\n", argv[0]);
```
fixed

danielinux commented on examples/dtls-wolfssl/dtls-server.c:
```diff
@@ -0,0 +1,135 @@
+#define SERVER_PORT 11111
+#define DEBUG 1
+extern const unsigned char server_cert[788];
+extern const unsigned char server_key[121];
```
fixed

danielinux commented:
> > Stack (or more likely heap) might be a problem initializing large objects (TLS session) on some >platforms. Wolfcrypt-only tests have more chances to run on those.
> 
> Ok. IIRC, wolfssl also has some option to work with static memory - isn't that used for RIOT? If it is used, I'd actually expect that the build would fail with some "region whatever overflowed" message for too constrained devices. If not, what needs to be done to use static memory?

Static memory 'moves' the problem of using the heap by allocating large objects in the stack, and it's genereally more memory consuming than normal allocations, because it also pre-allocates large pools of static memory to be able to be flexible enough for a generic use case. I have asked @kaleb-himes and he recommends to keep WOLFSSL_STATIC off in memory constrained environments.

MichelRottleuthner commented:
Makes sense. I didn't know how variable the size of the buffers are that wolfssl needs (and by that how bad the overhead of static memory would be). That also means that we have to either exclude all platforms that have no proper implementation of malloc available (e.g. only one-way-malloc on MSP430, not sure about others) or as an alternative we use the tlsf based malloc implementation there. Adding support for tlsf can be easily done later in a follow-up though.

danielinux commented on examples/dtls-wolfssl/README.md:
```diff
@@ -0,0 +1,43 @@
+
+This example is configured to use socks instead of sockets (over GNRC).
+It's possible to use sockets, which give a more similar approach to the
+UNIX version of wolfSSL.
```
Re-worded, so the examples now refer to each other.

danielinux commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ * Copyright (C) 2018 Inria
```
updated

danielinux commented on examples/dtls-wolfssl/dtls-server.c:
```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ * Copyright (C) 2018 Inria
```
done

danielinux commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,128 @@
+ * directory for more details.
+ */
+/*
+ * **** This file incorporates work covered by the following copyright and ****
```
correct, re-marked all contributed files to this repository

danielinux commented on examples/dtls-wolfssl/dtls-server.c:
```diff
@@ -0,0 +1,135 @@
+ * directory for more details.
+ */
+/*
+ * **** This file incorporates work covered by the following copyright and ****
```
done

danielinux commented:
Michael, I fixed most of the comments above (still working on the others).
I've reworked the posix tls examples, and unified wolfssl_server / wolfssl_client in a shell application (tls-wolfssl-posix) that now has:
 * netif_default
 * ecc256 keys used by default as PK, this should make it work with less resources
 * IPv6-only sockets, to be LoWPAN friendly

MichelRottleuthner commented:
Very nice progress!
I will do another round of testing with the new test applications (hopefully tomorrow)

miri64 commented:
@danielinux re you aware about #11909 (and other PRs around #10897)? It is based on the API you introduced here, but more generalized for other DTLS APIs. It would be nice, if we could streamline both efforts, so feedback there would always be welcome.

MichelRottleuthner commented:
I did a few runs of `tests/wolfssl-test` on different platforms with the following changes:
(note: please add these changes to the PR to make it work on MCU targets out of the box)
added to `tests/wolfssl-test/Makefile`:
```diff
diff --git a/tests/wolfssl-test/Makefile b/tests/wolfssl-test/Makefile
index f0dc0be06..45e894cb1 100644
--- a/tests/wolfssl-test/Makefile
+++ b/tests/wolfssl-test/Makefile
@@ -45,5 +45,8 @@ USEMODULE += wolfcrypt wolfcrypt-test wolfcrypt_sha512 wolfcrypt_curve25519  \
 # Comment the following line to disable full-benchmark test
 USEMODULE += wolfcrypt-benchmark
 
+ifneq ($(BOARD),native)
+  CFLAGS += -DBENCH_EMBEDDED
+endif
 
 include $(RIOTBASE)/Makefile.include
```
added to `pkg/wolfssl/Makefile.dep`:
```diff
diff --git a/pkg/wolfssl/Makefile.dep b/pkg/wolfssl/Makefile.dep
index 65fb0bbf3..2fa9e37c9 100644
--- a/pkg/wolfssl/Makefile.dep
+++ b/pkg/wolfssl/Makefile.dep
@@ -7,6 +7,8 @@ endif
 ifneq (,$(filter wolfcrypt-benchmark,$(USEMODULE)))
   USEMODULE += wolfcrypt
   USEMODULE += wolfcrypt_coding
+  USEMODULE += printf_float
+  USEMODULE += xtimer
 endif
```
With these changes  test+benchmark worked perfectly on most of the platforms I tested. The only problems I had were on `c2650-launchpad` and `nucleo-l152re` which both compiled fine but hard faulted.

`examples/tls-wolfssl-posix` always failed for me (on native and any MCU I tried) on the client-side with
```
ERROR: failed to connect, error 22
```

pokgak commented:
Is there a reason for not including preshared key support in this PR? I managed to add support for it by adding `WOLFSSL_STATIC_PSK` define to `user_settings.h` and replacing the certificate loading functions with those mentioned in this [page][1]. I tested it using `examples/dtls-wolfssl` and managed to send and received successfully.

[1]: https://wolfssl.zendesk.com/hc/en-us/articles/211682606-Using-Pre-Shared-Keys-PSK-with-wolfSSL

danielinux commented:
> I did a few runs of `tests/wolfssl-test` on different platforms with the following changes:
> (note: please add these changes to the PR to make it work on MCU targets out of the box)
> added to `tests/wolfssl-test/Makefile`:

Done, thanks

> With these changes test+benchmark worked perfectly on most of the platforms I tested. The only problems I had were on `c2650-launchpad` and `nucleo-l152re` which both compiled fine but hard faulted.
> 

Might be a problem with stack size? Tests/benchmarks can be demanding.

> `examples/tls-wolfssl-posix` always failed for me (on native and any MCU I tried) on the client-side with
> 
> ```
> ERROR: failed to connect, error 22
> ```

Sorry, I should have tested this more thoroughly. I'll ensure I fix all the examples and address the last two open comments.

Thanks again for your time!

pokgak commented on pkg/wolfssl/include/user_settings.h:
```diff
@@ -73,6 +73,7 @@ int strncasecmp(const char *s1, const char * s2, unsigned int sz);
 #define TFM_TIMING_RESISTANT
 #define USE_CERT_BUFFERS_2048
 #define NO_RC4
+#define WOLFSSL_STATIC_PSK
```
Sorry for not stating this clearly in my last comment. I actually meant something like below, so that you can choose to compile wolfSSL with or without pre-shared key support at compile time with `USEMODULE += wolfcrypt_psk` just like ECC.

```C
#undef WOLFSSL_STATIC_PSK
#ifdef MODULE_WOLFCRYPT_PSK
  #define WOLFSSL_STATIC_PSK
#endif
```

danielinux replied:
Fixed, added module `USEMODULE += wolfssl_psk` . Updated dtls makefile to show the possible options for DH/PK, and the PSK alternative.

MichelRottleuthner commented:
Oh I forgot to mention in my previous comment: on MCU targets `tests/wolfssl-test` actually reported a problem `RTC/Time not set!` in the `asn` testing step. This seems to be related to #9187. TLDR: wolfSSL uses `gettimeofday()` wich uses `xtimer_now64()` in case no RTC is present. But `xtimer_now64()` can actually start with zero on boot, letting the test fail. To circumvent this we could either strictly require RTC support or, preferably, add `FEATURES_OPTIONAL += periph_rtc` and a workaround for boards that don't have an RTC which waits at least a second before the test starts.

danielinux commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,128 @@
+    /* attempt to connect until the connection is successful */
+    do {
+        printf("connecting to server...\n");
+        if (sock_dtls_session_create(sk) < 0)
```
I have made a few tests. The fact is that if the client hello is lost for some reason, you have to rely on the dtls session timeout to try again. I've modified the example so the error codes are separate and it's more clear what's happening. The session is now destroyed/reopened after five recv() timeouts (with consequent ERROR_WANT_READ).

danielinux commented:
> 
> `examples/tls-wolfssl-posix` always failed for me (on native and any MCU I tried) on the client-side with
> 
> ```
> ERROR: failed to connect, error 22
> ```

Hello @MichelRottleuthner ,

Regarding the posix example:

I have tried different things, it seems that the `connect()` call fails because lwIP refuses to bind the TCP socket automatically.

I have tried as well adding an explicit `bind()` to the client example, but the error is the same.
It is possible that lwIP gets confused by the link-local address. Do you know how I can add another address to the netif (e.g. a site-local address with the same suffix)? 

Is there any other TCP examples using lwIP with posix sockets, or using multi-homed IPv6?

miri64 commented:
Just a note: the `sock_dtls` API at https://github.com/RIOT-OS/RIOT/pull/11909 is now merged.

pokgak commented:
Hi @kaleb-himes @danielinux,

Sorry that this PR keeps getting dragged on. Michel is currently in vacation and will not be available until next week and other mantainers are also busy preparing for the summit so I think it will be hard to get this merge before that. I'm not a mantainer so even though I can help with the review and testing, I cannot give the ACK (or at least it doesn't mean much).

Regarding DTLS sock API (#11909), it contains functions with the same name as what is used in this PR (e.g. `sock_dtls_create/destroy()`), so those needs to be renamed.

Onto the code itself, there are still minor issues with the [formatting and coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions) but I think this can be done later after Michel is finished with his review to avoid clutter.

I have not run any of the tests or the `tls-wolfssl-posix` example yet but I tested the `dtls-wolfssl` example again yesterday and with the changes from danielinux/RIOT#2, I managed to send and receive when using `wolfcrypt_ecc` and `wolfssl_psk` but using `wolfcrypt_ed25519` and `wolfcrypt_rsa` the server fails in `sock_dtls_create()` with the following error:

```
2019-08-28 11:39:11,086 - INFO # dtlss start
2019-08-28 11:39:11,091 - INFO # Error allocating ssl session
2019-08-28 11:39:11,095 - INFO # Failed to create DTLS session (err: Not enough space)
```

I also tested this on a board with bigger memory (`nucleo-l476rg`) and setting a bigger `THREAD_STACKSIZE_MAIN` but I still got the same error. The server fails with the same error on `native`.

> Is there any other TCP example using lwIP with posix sockets, or using multi-homed IPv6?

@miri64 do you know any?

miri64 commented:
> > Is there any other TCP example using lwIP with posix sockets, or using multi-homed IPv6?
> 
> @miri64 do you know any?

No sorry :(

danielinux commented:
> Regarding DTLS sock API (#11909), it contains functions with the same name as what is used in this PR (e.g. `sock_dtls_create/destroy()`), so those needs to be renamed.

Why is this necessary? I thought that #11909 was unifying the DTLS layer to use the same function names as this PR. Isn't the name the same on purpose (i.e. different TLS library, same names?). Please clarify.

> 
> Onto the code itself, there are still minor issues with the [formatting and coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions) but I think this can be done later after Michel is finished with his review to avoid clutter.
> 

Yes, I believe there is a coding style checker in travis, when we are done with reviews I'll go through those warnings and squash the entire tree

> I have not run any of the tests or the `tls-wolfssl-posix` example yet but I tested the `dtls-wolfssl` example again yesterday and with the changes from [danielinux#2](https://github.com/danielinux/RIOT/pull/2), I managed to send and receive when using `wolfcrypt_ecc` and `wolfssl_psk` but using `wolfcrypt_ed25519` and `wolfcrypt_rsa` the server fails in `sock_dtls_create()` with the following error:
> 
> ```
> 2019-08-28 11:39:11,086 - INFO # dtlss start
> 2019-08-28 11:39:11,091 - INFO # Error allocating ssl session
> 2019-08-28 11:39:11,095 - INFO # Failed to create DTLS session (err: Not enough space)
> ```

I am looking into this now

> 
> I also tested this on a board with bigger memory (`nucleo-l476rg`) and setting a bigger `THREAD_STACKSIZE_MAIN` but I still got the same error. The server fails with the same error on `native`.
> 
> > Is there any other TCP example using lwIP with posix sockets, or using multi-homed IPv6?
> 
> @miri64 do you know any?

Since there is no indication on how to use lwIP with IPv6 sockets, I'll revert the tls-posix-example to use IPv4 only.

miri64 commented:
> Why is this necessary? I thought that #11909 was unifying the DTLS layer to use the same function names as this PR. Isn't the name the same on purpose (i.e. different TLS library, same names?). Please clarify.

The API in this PR sometimes uses different parameters and does not employ `credman`. The idea behind the API introduced in #11909 is not only to have common names, but to have a common interface to all DTLS implementations, so a user can easily switch between different implementations without any changes to the C code.

miri64 commented:
The ideal situation would be, that e.g. `gcoap` can use either `tinydtls`, `wolfssl` or whatever for `coaps://` resources without much complicated `ifdef`ing in the `gcoap` code-base.

pokgak commented:
> > Regarding DTLS sock API (#11909), it contains functions with the same name as what is used in this PR (e.g. `sock_dtls_create/destroy()`), so those needs to be renamed.
> 
> Why is this necessary? I thought that #11909 was unifying the DTLS layer to use the same function names as this PR. Isn't the name the same on purpose (i.e. different TLS library, same names?). Please clarify.

#11909 is a DTLS library-independent API *in addition* to the DTLS API provided by DTLS library (tinydtls/wolfSSL). It is meant to avoid using DTLS library-specific calls in applications like RIOT's own CoAP implementation - gcoap, so that it is not tied to any specific library. This means there will be two ways of using DTLS in RIOT: 1) using the API offered by the DTLS library directly and 2) using the DTLS sock API for a more generic approach. This is mentioned [here](http://api.riot-os.org/group__net__dtls.html).

Even though some of the functions have the same names, the behaviour described is not the same. For integration of wolfSSL into RIOT, it is enough for this PR to target the first way (using API of DTLS library directly).

Currently, this PR mixes calls to `sock_dtls_*` wrapper and API of wolfSSL. I, personally, find that it is better to use the wolfSSL API directly so that it is inline with what is done in the tinydtls example but this is just a minor issue and does not affect the logic used so that's why I haven't mentioned this before.

I hope this clear the confusion.

danielinux commented:
> I have not run any of the tests or the `tls-wolfssl-posix` example yet but I tested the `dtls-wolfssl` example again yesterday and with the changes from [danielinux#2](https://github.com/danielinux/RIOT/pull/2), I managed to send and receive when using `wolfcrypt_ecc` and `wolfssl_psk` but using `wolfcrypt_ed25519` and `wolfcrypt_rsa` the server fails in `sock_dtls_create()` with the following error:

In 1af476898e48d59a2a41fbfcde98176e81f5476b I fixed the configuration for RSA ciphersuites so now the dtls example works with `wolfssl_rsa`, and removed the ed25519 option.

pokgak commented:
Retested `dtls-wolfssl` example again, the results are as follows:

| board             | ECC           | RSA           | PSK         |
|------------------ |----------------|----------------|--------------|
| native             | PASSED    | PASSED    | PASSED |
| samr21-xpro | FAILED (1) | FAILED (2) | PASSED |

1. Server takes too long to respond, client thought server are not responding. Setting a longer timeout using `wolfSSL_set_timeout_init()` on the client side fixed this.
2. Server/Client: context before hardfault, Adding 512 bytes more in to `THREAD_STACKSIZE_MAIN` fixed this:
```
CFLAGS += -DTHREAD_STACKSIZE_MAIN=\(3*THREAD_STACKSIZE_DEFAULT+512\)
```
But client got stuck in the `wolfSSL_accept()` loop after `HelloVerifyRequest` from the server. Using `wolfSSL_ERR_error_string()`, I got error code -125, "out of memory error". Even with a longer timeout, the error is still there.

For ECC and PSK, the client/server also get `Error allocating ssl session` after the server is started a few times, or the client made a few connection.

Overall, the code logic works, but due to longer response time and memory issues on hardware the handshake failed.

danielinux commented:
@MichelRottleuthner @miri64 

regarding POSIX examples: in 96aa9e9ea2fb776a15814e3d6b450dbf1c5dedef I managed to set up a site-local address for both endpoints, that shows up with the `ifconfig` shell command now.
TCP connect/accept seems to work reasonably now that the IP addresses are homed.

**but**: I found (at least) two problems in the POSIX socket interface:

1) `accept()` does not assign the `sock` struct member properly. I fixed applying the following patch:

```diff
--- a/sys/posix/sockets/posix_sockets.c
+++ b/sys/posix/sockets/posix_sockets.c
@@ -461,6 +461,7 @@ int accept(int socket, struct sockaddr *restrict address,
                 new_s->bound = true;
                 new_s->queue_array = NULL;
                 new_s->queue_array_len = 0;
+                new_s->sock = (socket_sock_t *)sock;
                 memset(&s->local, 0, sizeof(sock_tcp_ep_t));
             }
             break;
```

Please consider including this fix in posix/sockets/.

2) `send[to]()` does not behave as expected, and sends truncated/malformed payloads. I suspect concurrent access to tcp output buffer, or something similar, because if I put a breakpoint in sendto() on both endpoints and hit continue to slow down the TLS handshake, the test completes successfully. It might be a problem with the native tap driver, I am still investigating this.

danielinux commented:
> Retested `dtls-wolfssl` example again, the results are as follows:
> board 	ECC 	RSA 	PSK
> native 	PASSED 	PASSED 	PASSED
> samr21-xpro 	FAILED (1) 	FAILED (2) 	PASSED

Hey @pokgak Thanks for testing this!!

> 
>     1. Server takes too long to respond, client thought server are not responding. Setting a longer timeout using `wolfSSL_set_timeout_init()` on the client side fixed this.


What is a value you would recommend as default for the example?

> 
>     2. Server/Client: context before hardfault, Adding 512 bytes more to `THREAD_STACKSIZE_MAIN` in the Makefile fixed this:
> 
> 
> ```
> CFLAGS += -DTHREAD_STACKSIZE_MAIN=\(3*THREAD_STACKSIZE_DEFAULT+512\)
> ```
> 
> But client got stuck in the `wolfSSL_accept()` loop after `HelloVerifyRequest` from the server. Using `wolfSSL_ERR_error_string()`, I got error code -125, "out of memory error". Even with a longer timeout, the error is still there.

RSA2048 is indeed perhaps on the side of asking too much for this target's resources

> 
> For ECC and PSK, the client/server also get `Error allocating ssl session` after the server is started a few times, or the client made a few connection. It seems to me like `sock_dtls_session_destroy()` doesn't actually frees the allocated memory like it is supposed to.

 * [ ] Thanks, noted. I am looking into this. 

> 
> Overall, the code logic works, but due to longer response time and memory issues on hardware the handshake failed.

These are inevitable. The examples are there to show some of the possibilites of the library. Hardware limitations are target specific, running the tests/benchmarks can help with the selection of the proper ciphersuite and [pseudo]module options for optimizations tailored to specific system and use cases.

miri64 commented:
>     1. `accept()` does not assign the `sock` struct member properly. I fixed applying the following patch:
> 
> 
> ```diff
> --- a/sys/posix/sockets/posix_sockets.c
> +++ b/sys/posix/sockets/posix_sockets.c
> @@ -461,6 +461,7 @@ int accept(int socket, struct sockaddr *restrict address,
>                  new_s->bound = true;
>                  new_s->queue_array = NULL;
>                  new_s->queue_array_len = 0;
> +                new_s->sock = (socket_sock_t *)sock;
>                  memset(&s->local, 0, sizeof(sock_tcp_ep_t));
>              }
>              break;
> ```
> 
> Please consider including this fix in posix/sockets/.

Done: https://github.com/RIOT-OS/RIOT/pull/12130. As the patch is yours I made you the author of the commit.
 
>     1. `send[to]()` does not behave as expected, and sends truncated/malformed payloads. I suspect concurrent access to tcp output buffer, or something similar, because if I put a breakpoint in sendto() on both endpoints and hit continue to slow down the TLS handshake, the test completes successfully. It might be a problem with the native tap driver, I am still investigating this.

If your PR is not based on current master, could you check if #11575 or #11576 fix the issue for you?

pokgak commented:
> What is a value you would recommend as default for the example?

Based on my testing 2 seconds is enough for the server to responds before the client gave up on the handshake (`connect_timeout == max_connect_timeouts`) and try to start a new one. With 2 seconds, the client will timeout a few times but it is lower than `max_connect_timeouts`. If no timeout at all is desired, then the server will need at least 4 seconds to responds.

> RSA2048 is indeed perhaps on the side of asking too much for this target's resources

I'll try to find a target with more resources to test this.

MichelRottleuthner commented on examples/tls-wolfssl-posix/wolfssl-server.c:
```diff
@@ -0,0 +1,187 @@
+ *
+ * @}
+ */
+
```
please add the following here so that strnlen is provided on all platforms:
```c
/* Required for strnlen in string.h, when building with -std=c99 */
#define _DEFAULT_SOURCE 1 
```

MichelRottleuthner commented on tests/wolfcrypt-ed25519-verify/Makefile:
```diff
@@ -0,0 +1,33 @@
+
+# Change this to 0 show compiler invocation lines by default:
+QUIET ?= 1
+
```
please add the following to avoid a hardfault on small platforms (e.g. amr21-xpro) after finishing the signature verification is complete:
```
CFLAGS += -DTHREAD_STACKSIZE_MAIN=THREAD_STACKSIZE_LARGE
```

MichelRottleuthner commented on tests/wolfssl-test/Makefile:
```diff
@@ -0,0 +1,52 @@
+# gcc v 5.4.0 on 16.04 LTS
+# This is an optimized stack value based on testing, if you observe
+# a segmentation fault please increase this stack size.
+CFLAGS += -DTHREAD_STACKSIZE_MAIN=35000
```
please change this to the following to have a big enough (but platform specific) stacksize to make it work on embedded targets:
`CFLAGS += -DTHREAD_STACKSIZE_MAIN=2*THREAD_STACKSIZE_LARGE`

MichelRottleuthner commented on tests/wolfssl-test/main.c:
```diff
@@ -0,0 +1,40 @@
+int main(void)
+{
+    puts("wolfSSL Crypto Test!");
+    wolfcrypt_test(NULL);
```
please add `xtimer_sleep(1);` before calling `wolfcrypt_test` (and the required `#include "xtimer.h"` to the top) to workaround a failing test on platforms that don't have a synced RTC, as I explained in [one of my previous comments](https://github.com/RIOT-OS/RIOT/pull/10308#issuecomment-519430658).

pokgak commented on examples/dtls-wolfssl/dtls-client.c:
```diff
@@ -0,0 +1,188 @@
+    int iface;
+    char *addr_str;
+    int connect_timeout = 0;
+    const int max_connect_timeouts = 5;
```
@MichelRottleuthner did you increase the timeout with `wolfSSL_set_timeout_init()` when testing this? IIRC this problem is due to the server taking too long to respond. Increasing `max_connect_timeouts` will have the same effect as increasing the timeout too but I prefer increasing the timeout here to show that we are giving the server longer to respond.

danielinux commented:
I fixed the latest comments, and removed the tls-posix-example for now, which is problematic due to missing symbols and undefined problems in the sys/posix interface.

Looking forward to receiving a final pre-approval to squash and fix travis errors.

maribu commented on pkg/wolfssl/sock_tls/sock_tls.h:
```diff
@@ -0,0 +1,344 @@
+ * @return  0 on success.
+ * @return  -EINVAL, if @sock is null or the SSL context is not initialized yet.
+ * @return  -ENOMEM, if not enough resources can be provided for the session to be
+ *          created
```
You could use `@retval` here. (Comment only, not insisting.)

danielinux replied:
I'll ignore this for now, I am not seeing any major advantage of using `retval` instead of `return`, but I am no doxygen expert.

maribu commented on pkg/wolfssl/sock_tls/sock_tls.c:
```diff
@@ -0,0 +1,96 @@
+void sock_dtls_set_endpoint(sock_tls_t *sk, const sock_udp_ep_t *addr)
+{
+    printf("wolfSSL: Setting peer address and port\n");
+    memcpy(&sk->peer_addr, addr, sizeof (sock_udp_ep_t));
```
I don't get why here `memcpy()` is used and later `XMEMCPY()` is used. My gut feeling is that this is unintentional. (Remark, not insisting on changes.)

danielinux replied:
Well spotted

maribu commented on tests/pkg_wolfcrypt-ed25519-verify/tools/ed25519_sign_msg.c:
```diff
@@ -0,0 +1,78 @@
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/wolfcrypt/ed25519.h>
+
+void print_buf(uint8_t *buf, int len)
```
(Again, comment only): You could use module `ocd` here to safe some lines of code here

danielinux replied:
agree, done

danielinux commented:
I will probably have to blacklist all Atmel samX devices, due to a conflicting symbol, unless anyone has better ideas on this regard.

On these platforms build currently fails and looks like this:
```
In file included from ../wolfssl/internal.h:57:0,
                 from keys.c:33:
../wolfssl/wolfcrypt/aes.h:205:3: error: conflicting types for 'Aes'
 } Aes;
   ^~~
In file included from ../../../../cpu/sam0_common/include/vendor/saml21/include/saml21j18a.h:263:0,
                 from ../../../../cpu/sam0_common/include/vendor/sam0.h:40,
                 from ../../../../cpu/sam0_common/include/cpu_conf.h:30,
                 from ../../../../core/include/thread.h:125,
                 from ../../../../sys/include/net/gnrc/netapi.h:62,
                 from ../../../../sys/include/net/gnrc.h:291,
                 from ../../../../sys/net/gnrc/sock/include/sock_types.h:30,
                 from ../wolfssl/wolfio.h:496,
                 from ../wolfssl/ssl.h:134,
                 from ../wolfssl/internal.h:29,
                 from keys.c:33:
../../../../cpu/sam0_common/include/vendor/saml21/include/component/aes.h:342:3: note: previous declaration of 'Aes' was here
 } Aes;
   ^~~
```

kaspar030 commented:
> I will probably have to blacklist all Atmel samX devices, due to a conflicting symbol, unless anyone has better ideas on this regard.

Maybe a ```#define _SAML21_AES_COMPONENT_``` above the include list somewhere would blank out the sam0 part of the conflict?

danielinux commented:
> Maybe a `#define _SAML21_AES_COMPONENT_` above the include list somewhere would blank out the sam0 part of the conflict? 

it worked by defining both `_SAML21_AES_COMPONENT_` and `_SAME54_AES_COMPONENT_` for the two platforms. Thanks @kaspar030 !

danielinux commented:
@MichelRottleuthner : reporting in, :robot: :robot: happy. Is there anything else to address?

MichelRottleuthner commented:
@danielinux very nice! Please go ahead and squash the fixing commits so there are only a few cleanly separated commits left, e.g. one adding the package, and one for each test app etc.
I'll give it a final round of testing and code/documentation check tomorrow (I don't have the hardware at hand) and If everything is fine we're ready to go! :tada:

MichelRottleuthner commented on tests/pkg_wolfssl/Makefile:
```diff
@@ -0,0 +1,64 @@
+#LDFLAGS += -g
+
+# Necessary to set the stacksize on Ubuntu Native device when using
+# gcc v 5.4.0 on 16.04 LTS
```
this comment regarding ubuntu and compiler version should be removed as it is not really depending on that.

danielinux replied:
done

MichelRottleuthner commented on tests/pkg_wolfssl/main.c:
```diff
@@ -0,0 +1,46 @@
+ * @file
+ * @brief       wolfSSL cryptographic library test
+ *
+ * @author      Kaleb J. Himes <kaleb@wolfssl.com>
```
@danielinux please also add yourself as an author

danielinux replied:
done

danielinux commented on examples/dtls-wolfssl/Makefile:
```diff
@@ -0,0 +1,63 @@
+
+CFLAGS += -DDTLS_DEFAULT_PORT=$(DTLS_PORT) -DDTLS_WOLFSSL -Wno-unused-parameter -Wno-unused-variable
+
+# FIXME: This is a temporary patch taken from the TinyDTLS example to ensure proper stack allocation
```
done

danielinux commented on pkg/wolfssl/Makefile.dep:
```diff
@@ -0,0 +1,84 @@
+endif
+
+ifneq (,$(filter wolfcrypt_aes,$(USEMODULE)))
+    USEMODULE += wolfcrypt_cmac
```
done

danielinux commented on pkg/wolfssl/doc.txt:
```diff
@@ -0,0 +1,160 @@
+ * DTLS example over GNRC UDP/IP stack.
+ * See documentation in `examples/dtls-wolfssl/README.md`
+ *
+ * TLS Client and server Example
```
done

danielinux commented on pkg/wolfssl/include/user_settings.h:
```diff
@@ -0,0 +1,322 @@
+#undef WOLFSSL_SHA384
+#undef WOLFSSL_SHA512
+#ifdef MODULE_WOLFCRYPT_SHA512
+#   define HAVE_SHA384
```
done

danielinux commented:
@MichelRottleuthner Waiting for you to double check on the latest fixup. I will squash again if everything looks OK. Thanks!!

MichelRottleuthner commented:
Perfect! Enough annoying whitespace fixes ;) go ahead and squash

miri64 commented:
What is the plan for `sock_dtls` now? Will @pokgak port it to our version?

miri64 commented:
Congratulations on finally merging it btw :tada:

danielinux commented:
It was our pleasure!
Expect more contributions from us very soon :) :wolf:

MichelRottleuthner commented:
>What is the plan for sock_dtls now? Will @pokgak port it to our version?

I'd be fine with that obviously, but I don't care as long as it gets done eventually.

danielinux commented:
> What is the plan for `sock_dtls` now? Will @pokgak port it to our version?

for now the decision was to keep the symbols with the same names, but of course I am available to support @pokgak in the unification of the API

kaspar030 commented:
> It was our pleasure!

:laughing:

pokgak commented:
Congratulations for the merge!

> What is the plan for `sock_dtls` now? Will @pokgak port it to our version?

I plan to work on this soon:tm: :)

bergzand commented:
congrats @danielinux!

OlegHahm commented:
Wooohoooo! Great news! Thanks for the effort and patience.

emmanuelsearch commented:
yay ;)

kYc0o commented:
Congrats @danielinux !!! I love to see this in!
